File: examples/java/src/main/java/io/automaguard/example/Agent.java
// Customer Data Assistant — AutomaGuard Java SDK example.
//
// Demonstrates AutomaGuard policy enforcement on a Spring AI ChatClient via
// the AutomaGuardAdvisor integration.
//
// Two prompts are included:
//
//   --safe    Aggregate query, no PII accessed or sent externally.
//             All events receive Allow or Audit verdicts.
//
//   --unsafe  Adversarial prompt that attempts to exfiltrate a customer's
//             PII via an external email.  The NoPIIExfiltration invariant
//             and the unapproved-domain rule block the email before it sends.
//
// Usage:
//   mvn -q package -DskipTests
//   java -jar target/customer-data-assistant.jar --safe
//   java -jar target/customer-data-assistant.jar --unsafe

package io.automaguard.example;

import io.automaguard.EnforcementException;
import io.automaguard.springai.AutomaGuardAdvisor;
import org.springframework.ai.chat.client.ChatClient;
import org.springframework.ai.openai.OpenAiChatModel;
import org.springframework.ai.openai.OpenAiChatOptions;
import org.springframework.ai.openai.api.OpenAiApi;

import java.nio.file.Files;
import java.nio.file.Path;
import java.nio.file.Paths;
import java.util.Arrays;

public final class Agent {

    public static void main(String[] args) throws Exception {
        // ── Policy path ───────────────────────────────────────────────────────
        Path policyPath = Paths.get(
                System.getProperty("user.dir"),
                "../customer_data_guard.aegisc").toAbsolutePath().normalize();

        if (!Files.exists(policyPath)) {
            System.err.println("Policy bytecode not found at " + policyPath);
            System.err.println("Compile it first:");
            System.err.println("  aegisc compile examples/customer_data_guard.aegis" +
                               " -o examples/customer_data_guard.aegisc");
            System.exit(1);
        }

        // ── Mode selection ────────────────────────────────────────────────────
        boolean safe   = Arrays.asList(args).contains("--safe");
        String  mode   = safe ? "safe" : "unsafe";
        String  prompt = safe ? Prompts.SAFE : Prompts.UNSAFE;

        System.out.println("\n=== AutomaGuard Java Example (" + mode + " run) ===\n");
        System.out.println("Prompt: " + prompt + "\n");

        // ── API key ───────────────────────────────────────────────────────────
        String apiKey = System.getenv("OPENAI_API_KEY");
        if (apiKey == null || apiKey.isBlank()) {
            System.err.println("Error: OPENAI_API_KEY environment variable is not set.");
            System.exit(1);
        }

        // ── Spring AI setup ───────────────────────────────────────────────────
        var openAiApi = new OpenAiApi(apiKey);
        var chatModel = new OpenAiChatModel(openAiApi,
                OpenAiChatOptions.builder().model("gpt-4o").build());

        var advisor = new AutomaGuardAdvisor(
                policyPath.toString(),
                (result, toolName) -> System.out.println(
                        "  [audit] " + toolName + ": " + result.getReason() +
                        " (rules: " + result.getTriggeredRules() + ")")
        );

        var client = ChatClient.builder(chatModel)
                .defaultAdvisors(advisor)
                .defaultTools(new CustomerDataTools())
                .build();

        // ── Run ───────────────────────────────────────────────────────────────
        try {
            String result = client.prompt(prompt).call().content();
            System.out.println("\nResult: " + result);
            System.exit(0);
        } catch (EnforcementException ex) {
            System.err.println("\nBLOCKED by AutomaGuard policy:");
            System.err.println("  Reason:  " + ex.getMessage());
            System.err.println("  Verdict: " + ex.getResult().getVerdict());
            if (!ex.getResult().getTriggeredRules().isEmpty()) {
                System.err.println("  Rules:   " + ex.getResult().getTriggeredRules());
            }
            ex.getResult().getViolations().forEach(v ->
                System.err.println("  Invariant: " + v));
            ex.getResult().getConstraintViolations().forEach(cv ->
                System.err.println("  Constraint: " + cv));
            System.exit(1);
        } finally {
            advisor.close();
        }
    }
}

File: examples/dotnet/Agent.cs
// Customer Data Assistant — AutomaGuard .NET SDK example.
//
// Demonstrates AutomaGuard policy enforcement on a Microsoft Semantic Kernel
// agent via the IFunctionInvocationFilter integration.
//
// Two prompts are included:
//
//   --safe    Aggregate query, no PII accessed or sent externally.
//             All events receive Allow or Audit verdicts.
//
//   --unsafe  Adversarial prompt that attempts to exfiltrate a customer's
//             PII via an external email.  The NoPIIExfiltration invariant
//             and the unapproved-domain rule block the email before it sends.
//
// Usage:
//   dotnet run -- --safe
//   dotnet run -- --unsafe

using AutomaGuard;
using AutomaGuard.SemanticKernel;
using AutomaGuard.Example;
using Microsoft.SemanticKernel;
using Microsoft.SemanticKernel.Connectors.OpenAI;

// ── Policy path ───────────────────────────────────────────────────────────────

// Walk up from the build output directory to examples/customer_data_guard.aegisc
var policyPath = Path.GetFullPath(
    Path.Combine(AppContext.BaseDirectory, "..", "customer_data_guard.aegisc"));

if (!File.Exists(policyPath))
{
    Console.Error.WriteLine($"Policy bytecode not found at {policyPath}");
    Console.Error.WriteLine("Compile it first:");
    Console.Error.WriteLine(
        "  aegisc compile examples/customer_data_guard.aegis " +
        "-o examples/customer_data_guard.aegisc");
    return 1;
}

// ── Mode selection ────────────────────────────────────────────────────────────

var mode   = args.Contains("--safe") ? "safe" : "unsafe";
var prompt = args.Contains("--safe") ? Prompts.Safe : Prompts.Unsafe;

Console.WriteLine($"\n=== AutomaGuard .NET Example ({mode} run) ===\n");
Console.WriteLine($"Prompt: {prompt}\n");

// ── Kernel setup ──────────────────────────────────────────────────────────────

var apiKey = Environment.GetEnvironmentVariable("OPENAI_API_KEY")
    ?? throw new InvalidOperationException(
        "OPENAI_API_KEY environment variable is not set.");

// AddAutomaGuard registers an IFunctionInvocationFilter that:
//   - Before each tool call: emits a tool_call / data_access / external_request
//     event (inferred from the function name and parameters) and throws
//     EnforcementException if the verdict is Deny.
//   - After each tool call: evaluates the result for redaction opportunities.
var kernel = Kernel.CreateBuilder()
    .AddOpenAIChatCompletion("gpt-4o", apiKey)
    .AddAutomaGuard(policyPath, options =>
    {
        options.OnAudit = (result, functionName) =>
            Console.WriteLine(
                $"  [audit] {functionName}: {result.Reason ?? "no reason"} " +
                $"(rules: [{string.Join(", ", result.TriggeredRules)}])");
    })
    .Build();

kernel.ImportPluginFromType<CustomerDataPlugin>();

var settings = new OpenAIPromptExecutionSettings
{
    ToolCallBehavior = ToolCallBehavior.AutoInvokeKernelFunctions,
};

// ── Run ───────────────────────────────────────────────────────────────────────

try
{
    var result = await kernel.InvokePromptAsync(
        prompt,
        new KernelArguments(settings));

    Console.WriteLine($"\nResult: {result}");
    return 0;
}
catch (EnforcementException ex)
{
    Console.Error.WriteLine("\nBLOCKED by AutomaGuard policy:");
    Console.Error.WriteLine($"  Reason:  {ex.Message}");
    Console.Error.WriteLine($"  Verdict: {ex.Result.Verdict}");

    if (ex.Result.TriggeredRules.Length > 0)
        Console.Error.WriteLine(
            $"  Rules:   [{string.Join(", ", ex.Result.TriggeredRules)}]");

    foreach (var v in ex.Result.Violations)
        Console.Error.WriteLine(
            $"  Invariant: {v.ProofName}/{v.InvariantName}: {v.Message}");

    return 1;
}
